File: ch02/2.4.4_binary_search/binary_search.py
#!/usr/bin/env python
#
# Copyright (C) 2017 ShadowMan
#


class BinarySearch(object):

    def __init__(self, sequence):
        self._sequence = sequence

    def search(self, element):
        start = 0
        end = len(self._sequence) - 1

        while start <= end:
            center = (start + end) // 2

            if element > self._sequence[center]:
                start = center + 1
            elif element < self._sequence[center]:
                end = center - 1
            else:
                return center
        return -1


if __name__ == '__main__':
    searcher = BinarySearch([0, 1, 3, 5, 6, 7, 8, 9])

    print("Binary Search: {}".format(searcher.search(2)))
    print("Binary Search: {}".format(searcher.search(5)))
    print("Binary Search: {}".format(searcher.search(9)))

File: ch02/2.4.4_binary_search/binary_search.c
#include <stdio.h>
#define NOT_FOUND -1

typedef int position;

position binary_search(const int element, const int sequence[], const size_t sequence_size);


int main(int argc, char *argv[]) {
    int sequence[] = { 0, 2, 3, 5, 6, 7, 9 };

    printf("Binary Search: %d\n", binary_search(1, sequence, sizeof(sequence) / sizeof(sequence[0])));
    printf("Binary Search: %d\n", binary_search(3, sequence, sizeof(sequence) / sizeof(sequence[0])));
    printf("Binary Search: %d\n", binary_search(6, sequence, sizeof(sequence) / sizeof(sequence[0])));
    return 0;
}


position binary_search(const int element, const int sequence[], const size_t sequence_size) {
    position start = 0, end = sequence_size - 1;
    position center;

    while (start <= end) {
        //        ^ start may be equal to end(Found)
        center = (start + end) / 2;

        if (element > sequence[center]) {
            start = center + 1;
        } else if (element < sequence[center]) {
            end = center - 1;
        } else {
            return center;
        }
    }
    return NOT_FOUND;
}
